modules/siddhi-core/src/main/java/org/wso2/siddhi/core/aggregation/AggregationRuntime.java

```diff
@@ -157,62 +198,95 @@ public StreamEvent find(StateEvent matchingEvent, CompiledCondition compiledCond
         }
 
         // Create within expression
         Expression withinExpression;
+        Expression start = Expression.variable(additionalAttributes.get(0).getName());
+        Expression end = Expression.variable(additionalAttributes.get(1).getName());
+        Expression compareWithStartTime = Compare.compare(start, Compare.Operator.LESS_THAN_EQUAL,
+                Expression.variable("_TIMESTAMP"));
+        Expression compareWithEndTime = Compare.compare(Expression.variable("_TIMESTAMP"), Compare.Operator.LESS_THAN,
+                end);
+        withinExpression = Expression.and(compareWithStartTime, compareWithEndTime);
+
+        // Create start and end time expression
+        Expression startEndTimeExpression;
         if (within.getTimeRange().size() == 1) {
-            withinExpression = new AttributeFunction("incrementalAggregator", "within",
-                    within.getTimeRange().get(0),
-                    Expression.variable("_TIMESTAMP"));
+            startEndTimeExpression = new AttributeFunction("incrementalAggregator", "startTimeEndTime",
+                    within.getTimeRange().get(0));
         } else { // within.getTimeRange().size() == 2
-            withinExpression = new AttributeFunction("incrementalAggregator", "within",
-                    within.getTimeRange().get(0),
-                    within.getTimeRange().get(1), Expression.variable("_TIMESTAMP"));
+            startEndTimeExpression = new AttributeFunction("incrementalAggregator", "startTimeEndTime",
+                    within.getTimeRange().get(0), within.getTimeRange().get(1));
         }
 
-        // Get table definition. Table definitions for all the tables used to persist aggregates are similar.
-        // Therefore it's enough to get the definition from one table.
-        AbstractDefinition tableDefinition = ((Table) aggregationTables.values().toArray()[0]).getTableDefinition();
+        startTimeEndTimeExpressionExecutor = ExpressionParser.parseExpression(startEndTimeExpression,
+                matchingMetaInfoHolder.getMetaStateEvent(), matchingMetaInfoHolder.getCurrentState(), tableMap,
+                variableExpressionExecutors, siddhiAppContext, false, 0, queryName);
 
         // Create compile condition per each table used to persist aggregates.
         // These compile conditions are used to check whether the aggregates in tables are within the given duration.
         for (Map.Entry<TimePeriod.Duration, Table> entry : aggregationTables.entrySet()) {
             CompiledCondition withinTableCompileCondition = entry.getValue().compileCondition(withinExpression,
-                    aggregationTableMetaInfoHolder(matchingMetaInfoHolder, tableDefinition), siddhiAppContext,
-                    variableExpressionExecutors, tableMap, queryName);
+                    newStreamTableMetaInfoHolder, siddhiAppContext, variableExpressionExecutors, tableMap, queryName);
```

Comment:
Can you use some useful names then using "new" as in newStreamTableMetaInfoHolder. use something like StreamTableMetaInfoHolderWithStartEnd

Reply:
Done